deepchecks/tabular/context.py

```diff
@@ -221,16 +223,35 @@ def __init__(
                 template='For more information please refer to the Supported Models guide {link}')
             raise DeepchecksValueError(f'Received unsorted model_classes. {supported_models_link}')
 
-        self._task_type, self._observed_classes, self._model_classes = infer_task_type_and_classes(
-            model, train, test, y_pred_train, y_pred_test, model_classes)
-
-        if model is None and \
-                not pd.Series([y_pred_train, y_pred_test, y_proba_train, y_proba_test]).isna().all():
+        if model_classes is None:
+            model_classes = infer_classes_from_model(model)
+        labels = None
+        if task_type is None:
+            labels = get_all_labels(model, train, test, y_pred_train, y_pred_test)
+            task_type = infer_task_type(train, labels, model_classes)
+
+        need_observed_classes_for_dummy_model = \
```

Comment:
why not place lines 233 up to 241 right before the dummy model creation? could simplify this boolean variable a bit